Floor Icons 2: The Return

NanakoAC opened the pull request:
Second stage of requested floor icon work
![maintfloor](https://i.imgur.com/Z7klPWm.png)
This PR fully implements support for whitelists, blacklists, and the greylist (which is just using both lists at once).
This allows specific floor types to be used as exceptions to the smoothing rule

But more importantly it implements atom based smoothing
This has a flexible, if rather complex system, which matches atoms by type, and also by exact values of variables, as well as a priority value to govern how values in opposite lists override each other

Included is an example, implemented on plating, to make a floor not smooth with dense anchored objects, except airlocks

LethalGhost commented on code/game/turfs/flooring/flooring.dm:
```diff
@@ -55,16 +55,60 @@ var/list/flooring_types
 	//How we smooth with other flooring
 	var/floor_smooth = SMOOTH_ALL
+	var/list/flooring_whitelist = list() //Smooth with nothing except the contents of this list
+	var/list/flooring_blacklist = list() //Smooth with nothing except the contents of this list
```
Wrong comment

LethalGhost replied:
But it still same as comment for whitelist. That's not right.

LethalGhost commented on code/game/turfs/simulated/floor_icon.dm:
```diff
@@ -132,28 +140,171 @@ var/list/flooring_cache = list()
+						var/match = TRUE
+						for (var/d_var in d_vars)
+							//For each variable we want to check
+							if (AM.vars[d_var] != d_vars[d_var])
```
`AM.vars[d_var]` will runtime if `d_var` not in `AM.vars`

NanakoAC replied:
I'm aware, the comment indicates as such.
But it will be in there, because the type is already checked beforehand.

LethalGhost replied:
That's only part of problem.
That specified type can be change but changer will have no clue that it vars uses here.

LethalGhost commented on code/game/turfs/flooring/flooring.dm:
```diff
@@ -55,16 +55,60 @@ var/list/flooring_types
+	smooth_movable_atom = SMOOTH_GREYLIST
+	movable_atom_blacklist = list(
+		list(/obj, list("density" = TRUE, "anchored" = TRUE), 1)
+		)
```
Wouldn't it be easier if items order will define priority?

NanakoAC replied:
sadly not, because the priority is primarily to determine prominence across both lists, not just within a single list

LethalGhost replied:
But can't we merge lists then?
Remove most `SMOOTH_X` defines and update list strucure that way:
```
list_name = list(
	list(/obj, list(vars), BLACKLIST / WHITELIST)
)
```

Priority will be handled with oreder.
That also help remove copypasted code below.

NanakoAC replied:
ehh, i guess. maybe it'd be better that way
I really just want to get this merged and done with though

LethalGhost commented on code/game/turfs/simulated/floor_icon.dm:
```diff
@@ -132,28 +140,171 @@ var/list/flooring_cache = list()
+
+
+		if (smooth_movable_atom == SMOOTH_BLACKLIST || smooth_movable_atom == SMOOTH_GREYLIST)
+			//All of this blacklist code is copypasted from above, with only minor name changes
```
>code is copypasted from above

Ugh

NanakoAC replied:
got an alternative idea?

NanakoAC commented on code/game/turfs/simulated/floor_icon.dm:
```diff
@@ -132,28 +140,171 @@ var/list/flooring_cache = list()
+							break
+	else
+		//If this is a countercheck, we skip all of the above, start off with true, and go straight to the atom lists
+		is_linked = TRUE
```
why do that? 
it doesn't really make much difference either way

LethalGhost replied:
Yeah but that structure will be easy to read and update. 
Structures like that is not good style and here it can be easily avoided:
```
if(condition 1)
	if(condition 2)
		if(condition 3)
			if(condition 4)
				if(condition 5)
					if(condition 6)
						if(condition 7)
							if(condition 8)
								return
							else
								expression 8
						else
							expression 7
					else
						expression 6
				else
					expression 5
			else
				expression 4
		else
			expression 3
	else
		expression 2
else
	expression 1

```

NanakoAC commented:
New Commit up, fixes several requested changes and conflicts, rebases to master. 
Also fixed several visual bugs with grass and simplified external edge code. Now grass turfs have external corners, and that bug with lines on dirt is gone

LethalGhost commented on code/game/turfs/simulated/floor_icon.dm:
```diff
@@ -132,28 +135,174 @@ var/list/flooring_cache = list()
+
+		if (smooth_movable_atom == SMOOTH_BLACKLIST || smooth_movable_atom == SMOOTH_GREYLIST)
+			//All of this blacklist code is copypasted from above, with only minor name changes
+			if (smooth_movable_atom == SMOOTH_BLACKLIST || smooth_movable_atom == SMOOTH_GREYLIST)
```
That check is redundand.
![default](https://user-images.githubusercontent.com/2458675/48902591-d1ea1580-ee7a-11e8-9f1b-55038f4a6568.PNG)
